File: nxflix-agents-ts/src/agents/study-orchestrator.ts
import { z } from 'zod';
import { v4 as uuidv4 } from 'uuid';
import { LLMProvider } from '../providers/llm.js';
import { SM2Service } from '../services/spaced-repetition.js';
import { GrammarService } from '../services/grammar.js';
import { createTrace } from '../tracing/index.js';
import type {
  StudyRecommendation,
  RecommendationRequest,
  SessionRequest,
} from '../models/study.js';
import type { UserProgress, StudySession, SessionResult } from '../models/progress.js';
import type { GrammarPoint } from '../models/grammar.js';

const RecommendationResponse = z.object({
  grammar_ids: z.array(z.string()),
  reason: z.string(),
  focus_areas: z.array(z.string()),
  suggested_question_types: z.array(z.string()),
});

export class StudyOrchestratorAgent {
  private grammarService: GrammarService;
  private sm2Service: SM2Service;
  private llm: LLMProvider;
  private sessions: Map<string, StudySession> = new Map();

  constructor(
    grammarService: GrammarService,
    sm2Service?: SM2Service,
    llmProvider?: LLMProvider
  ) {
    this.grammarService = grammarService;
    this.sm2Service = sm2Service ?? new SM2Service();
    this.llm = llmProvider ?? new LLMProvider();
  }

  async getRecommendations(
    request: RecommendationRequest,
    userProgress: UserProgress[]
  ): Promise<StudyRecommendation> {
    const trace = createTrace('study_orchestrator.get_recommendations', {
      userId: request.userId,
      maxItems: request.maxItems,
    });

    try {
      const dueItems = this.sm2Service.getDueItems(userProgress, request.maxItems);

      const availableGrammar = this.grammarService.getGrammarPointsByIds(
        request.availableGrammarIds
      );

      const progressSummary = this.buildProgressSummary(userProgress);
      const dueSummary = this.buildDueSummary(dueItems);
      const grammarSummary = this.buildGrammarSummary(availableGrammar.slice(0, 20));

      const prompt = `You are a JLPT N1 study advisor. Based on the student's progress, recommend grammar points to study.

## Student Progress Summary
${progressSummary}

## Items Due for Review
${dueSummary}

## Available Grammar Points (sample)
${grammarSummary}

## Request
- Maximum items: ${request.maxItems}
- Focus on weak areas: ${request.focusWeakAreas}
- Include new items: ${request.includeNew}
- Time available: ${request.timeAvailableMinutes ?? 'unlimited'} minutes

Select the most appropriate grammar points to study and explain why. Consider:
1. Items that are due for review (highest priority)
2. Items with low mastery levels
3. A mix of review and new content
4. Related grammar patterns that reinforce each other

Respond with a JSON object containing:
- grammar_ids: list of grammar IDs to study (up to ${request.maxItems})
- reason: brief explanation of your recommendation
- focus_areas: specific areas the student should focus on
- suggested_question_types: recommended question types for practice`;

      const response = await this.llm.completeJson(
        [{ role: 'user', content: prompt }],
        RecommendationResponse
      );

      // Validate recommended grammar IDs exist
      let validIds = response.grammar_ids.filter(
        (id) => this.grammarService.getGrammarPoint(id) !== undefined
      );

      // Fall back to due items if LLM recommendations are invalid
      if (validIds.length === 0 && dueItems.length > 0) {
        validIds = dueItems.slice(0, request.maxItems).map((p) => p.grammarId);
      }

      const recommendation: StudyRecommendation = {
        grammarIds: validIds,
        reason: response.reason,
        priority: dueItems.length > 0 ? 1 : 2,
        estimatedTimeMinutes: validIds.length * 3,
        focusAreas: response.focus_areas,
        suggestedQuestionTypes: response.suggested_question_types,
      };

      trace?.update({ output: { grammarCount: validIds.length } });
      trace?.end();

      return recommendation;
    } catch (error) {
      trace?.update({ output: { error: String(error) } });
      trace?.end();
      throw error;
    }
  }

  async startSession(request: SessionRequest): Promise<StudySession> {
    const trace = createTrace('study_orchestrator.start_session', {
      userId: request.userId,
      grammarCount: request.grammarIds.length,
    });

    const sessionId = uuidv4();

    const session: StudySession = {
      id: sessionId,
      userId: request.userId,
      grammarIds: request.grammarIds,
      startedAt: new Date().toISOString(),
      completedAt: null,
      results: [],
      totalQuestions: request.questionCount,
      totalCorrect: 0,
    };

    this.sessions.set(sessionId, session);

    trace?.update({ output: { sessionId } });
    trace?.end();

    return session;
  }

  async completeSession(
    sessionId: string,
    results: SessionResult[],
    userProgress: Record<string, UserProgress>
  ): Promise<{ session: StudySession; updatedProgress: Record<string, UserProgress> }> {
    const trace = createTrace('study_orchestrator.complete_session', { sessionId });

    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error(`Session not found: ${sessionId}`);
    }

    // Update session with results
    session.results = results;
    session.completedAt = new Date().toISOString();
    session.totalCorrect = results.reduce((sum, r) => sum + r.correctAnswers, 0);
    session.totalQuestions = results.reduce((sum, r) => sum + r.questionsAsked, 0);

    // Update user progress using SM-2
    const updatedProgress: Record<string, UserProgress> = {};
    for (const result of results) {
      if (result.grammarId in userProgress) {
        const progress = userProgress[result.grammarId];
        const quality = Math.round(result.score);
        updatedProgress[result.grammarId] = this.sm2Service.updateProgress(
          progress,
          quality
        );
      }
    }

    trace?.update({ output: { totalCorrect: session.totalCorrect } });
    trace?.end();

    return { session, updatedProgress };
  }

  getSession(sessionId: string): StudySession | undefined {
    return this.sessions.get(sessionId);
  }

  private buildProgressSummary(progress: UserProgress[]): string {
    if (progress.length === 0) {
      return 'No previous study history.';
    }

    const total = progress.length;
    const studied = progress.filter((p) => p.timesStudied > 0).length;
    const avgMastery = progress.reduce((sum, p) => sum + p.masteryLevel, 0) / total;

    const weakItems = progress.filter(
      (p) => p.masteryLevel < 3 && p.timesStudied > 0
    );

    let summary = `- Total grammar points: ${total}
- Previously studied: ${studied}
- Average mastery level: ${avgMastery.toFixed(1)}/5
- Items needing review: ${weakItems.length}`;

    if (weakItems.length > 0) {
      const weakIds = weakItems.slice(0, 5).map((p) => p.grammarId);
      summary += `\n- Weakest items: ${weakIds.join(', ')}`;
    }

    return summary;
  }

  private buildDueSummary(dueItems: UserProgress[]): string {
    if (dueItems.length === 0) {
      return 'No items currently due for review.';
    }

    return `- ${dueItems.length} items due for review
- Item IDs: ${dueItems.slice(0, 10).map((p) => p.grammarId).join(', ')}`;
  }

  private buildGrammarSummary(grammar: GrammarPoint[]): string {
    if (grammar.length === 0) {
      return 'No grammar points available.';
    }

    return grammar
      .slice(0, 10)
      .map((g) => `- ${g.id}: ${g.pattern} (${g.meaning})`)
      .join('\n');
  }
}

File: nxflix-agents-python/src/nxflix_agents/agents/study_orchestrator.py
"""Study Orchestrator Agent - Manages personalized study recommendations and sessions."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel

from nxflix_agents.models import (
    GrammarPoint,
    UserProgress,
    StudySession,
    SessionResult,
    StudyRecommendation,
    RecommendationRequest,
    SessionRequest,
)
from nxflix_agents.providers import LLMProvider
from nxflix_agents.services import SM2Service, GrammarService
from nxflix_agents.tracing.setup import track


class RecommendationResponse(BaseModel):
    """LLM response for study recommendations."""

    grammar_ids: list[str]
    reason: str
    focus_areas: list[str]
    suggested_question_types: list[str]


class StudyOrchestratorAgent:
    """Agent responsible for orchestrating study sessions and recommendations."""

    def __init__(
        self,
        grammar_service: GrammarService,
        sm2_service: SM2Service | None = None,
        llm_provider: LLMProvider | None = None,
    ):
        self.grammar_service = grammar_service
        self.sm2_service = sm2_service or SM2Service()
        self.llm = llm_provider or LLMProvider()
        self._sessions: dict[str, StudySession] = {}

    @track("study_orchestrator.get_recommendations")
    async def get_recommendations(
        self,
        request: RecommendationRequest,
        user_progress: list[UserProgress],
    ) -> StudyRecommendation:
        """Get personalized study recommendations based on user progress.

        Args:
            request: Recommendation request with user preferences
            user_progress: User's current progress on grammar points

        Returns:
            Personalized study recommendation
        """
        # Get due items
        due_items = self.sm2_service.get_due_items(user_progress, limit=request.max_items)

        # Build context for LLM
        available_grammar = self.grammar_service.get_grammar_points_by_ids(
            request.available_grammar_ids
        )

        progress_summary = self._build_progress_summary(user_progress)
        due_summary = self._build_due_summary(due_items)
        grammar_summary = self._build_grammar_summary(available_grammar[:20])

        prompt = f"""You are a JLPT N1 study advisor. Based on the student's progress, recommend grammar points to study.

## Student Progress Summary
{progress_summary}

## Items Due for Review
{due_summary}

## Available Grammar Points (sample)
{grammar_summary}

## Request
- Maximum items: {request.max_items}
- Focus on weak areas: {request.focus_weak_areas}
- Include new items: {request.include_new}
- Time available: {request.time_available_minutes or 'unlimited'} minutes

Select the most appropriate grammar points to study and explain why. Consider:
1. Items that are due for review (highest priority)
2. Items with low mastery levels
3. A mix of review and new content
4. Related grammar patterns that reinforce each other

Respond with a JSON object containing:
- grammar_ids: list of grammar IDs to study (up to {request.max_items})
- reason: brief explanation of your recommendation
- focus_areas: specific areas the student should focus on
- suggested_question_types: recommended question types for practice"""

        response = await self.llm.complete_json(
            messages=[{"role": "user", "content": prompt}],
            response_model=RecommendationResponse,
        )

        # Validate recommended grammar IDs exist
        valid_ids = [
            gid for gid in response.grammar_ids
            if self.grammar_service.get_grammar_point(gid) is not None
        ]

        # Fall back to due items if LLM recommendations are invalid
        if not valid_ids and due_items:
            valid_ids = [p.grammar_id for p in due_items[:request.max_items]]

        return StudyRecommendation(
            grammar_ids=valid_ids,
            reason=response.reason,
            priority=1 if due_items else 2,
            estimated_time_minutes=len(valid_ids) * 3,
            focus_areas=response.focus_areas,
            suggested_question_types=response.suggested_question_types,
        )

    @track("study_orchestrator.start_session")
    async def start_session(self, request: SessionRequest) -> StudySession:
        """Start a new study session.

        Args:
            request: Session request with grammar IDs and settings

        Returns:
            New study session
        """
        session_id = str(uuid.uuid4())

        session = StudySession(
            id=session_id,
            user_id=request.user_id,
            grammar_ids=request.grammar_ids,
            started_at=datetime.utcnow(),
            results=[],
            total_questions=request.question_count,
            total_correct=0,
        )

        self._sessions[session_id] = session
        return session

    @track("study_orchestrator.complete_session")
    async def complete_session(
        self,
        session_id: str,
        results: list[SessionResult],
        user_progress: dict[str, UserProgress],
    ) -> tuple[StudySession, dict[str, UserProgress]]:
        """Complete a study session and update user progress.

        Args:
            session_id: ID of the session to complete
            results: Results for each grammar point studied
            user_progress: Current progress for each grammar point

        Returns:
            Tuple of (completed session, updated progress)
        """
        session = self._sessions.get(session_id)
        if not session:
            raise ValueError(f"Session not found: {session_id}")

        # Update session with results
        session.results = results
        session.completed_at = datetime.utcnow()
        session.total_correct = sum(r.correct_answers for r in results)
        session.total_questions = sum(r.questions_asked for r in results)

        # Update user progress using SM-2
        updated_progress = {}
        for result in results:
            if result.grammar_id in user_progress:
                progress = user_progress[result.grammar_id]
                quality = self.sm2_service.quality_from_score(
                    result.score, max_score=5.0
                )
                updated_progress[result.grammar_id] = self.sm2_service.update_progress(
                    progress, quality
                )

        return session, updated_progress

    def get_session(self, session_id: str) -> StudySession | None:
        """Get a session by ID."""
        return self._sessions.get(session_id)

    def _build_progress_summary(self, progress: list[UserProgress]) -> str:
        """Build a text summary of user progress."""
        if not progress:
            return "No previous study history."

        total = len(progress)
        studied = sum(1 for p in progress if p.times_studied > 0)
        avg_mastery = sum(p.mastery_level for p in progress) / total if total else 0

        weak_items = [p for p in progress if p.mastery_level < 3 and p.times_studied > 0]

        summary = f"""- Total grammar points: {total}
- Previously studied: {studied}
- Average mastery level: {avg_mastery:.1f}/5
- Items needing review: {len(weak_items)}"""

        if weak_items:
            weak_ids = [p.grammar_id for p in weak_items[:5]]
            summary += f"\n- Weakest items: {', '.join(weak_ids)}"

        return summary

    def _build_due_summary(self, due_items: list[UserProgress]) -> str:
        """Build a text summary of items due for review."""
        if not due_items:
            return "No items currently due for review."

        return f"""- {len(due_items)} items due for review
- Item IDs: {', '.join(p.grammar_id for p in due_items[:10])}"""

    def _build_grammar_summary(self, grammar: list[GrammarPoint]) -> str:
        """Build a text summary of available grammar points."""
        if not grammar:
            return "No grammar points available."

        lines = []
        for g in grammar[:10]:
            lines.append(f"- {g.id}: {g.pattern} ({g.meaning})")

        return "\n".join(lines)
